src/hotspot/share/opto/memnode.cpp

```diff
@@ -2770,9 +2770,461 @@ Node *StoreNode::Ideal(PhaseGVN *phase, bool can_reshape) {
+          // LShiftL( ConvI2L(int_offset + int_con), int_offset_shift)
+          constant_offset += ((jlong)int_offset->in(2)->get_int()) << int_offset_shift;
+          int_offset = int_offset->in(1);
+        }
+      } else {
+        // All others
+        other_offsets->append(offset);
+      }
+    }
+
+    return ArrayPointer(true, pointer, base, constant_offset, int_offset, int_offset_shift, other_offsets);
+  }
+
+  bool is_adjacent_to(const ArrayPointer& other, const jlong data_size) const {
+    if (!_is_valid || !other._is_valid) { return false; }
+
+    // Offset adjacent?
+    if (this->_constant_offset + data_size != other._constant_offset) { return false; }
+
+    // All other components identical?
+    if (this->_base != other._base ||
+        this->_int_offset != other._int_offset ||
+        this->_int_offset_shift != other._int_offset_shift ||
+        this->_other_offsets->length() != other._other_offsets->length()) {
+      return false;
+    }
+
+    for (int i = 0; i < this->_other_offsets->length(); i++) {
+      Node* o1 = this->_other_offsets->at(i);
+      Node* o2 = other._other_offsets->at(i);
+      if (o1 != o2) { return false; }
+    }
+
+    return true;
+  }
+
+#ifndef PRODUCT
+  void dump() {
+    if (!_is_valid) {
+      tty->print("ArrayPointer[%d %s, invalid]", _pointer->_idx, _pointer->Name());
+      return;
+    }
+    tty->print("ArrayPointer[%d %s, base[%d %s] + %lld",
+               _pointer->_idx, _pointer->Name(),
+               _base->_idx, _base->Name(),
+               (long long)_constant_offset);
+    if (_int_offset != 0) {
+      tty->print(" + I2L[%d %s] << %d",
+                 _int_offset->_idx, _int_offset->Name(), _int_offset_shift);
+    }
+    for (int i = 0; i < _other_offsets->length(); i++) {
+      Node* n = _other_offsets->at(i);
+      tty->print(" + [%d %s]", n->_idx, n->Name());
+    }
+    tty->print_cr("]");
+  }
+#endif
+};
+
+StoreNode* StoreNode::can_merge_primitive_array_store_with_def(PhaseGVN* phase, bool check_use) {
+  int opc = Opcode();
+  assert(opc == Op_StoreB || opc == Op_StoreC || opc == Op_StoreI, "precondition");
+
+  StoreNode* def = in(MemNode::Memory)->isa_Store();
+  if (def == nullptr ||
+      def->Opcode() != opc ||
+      def->adr_type()->isa_aryptr() == nullptr) {
+    return nullptr;
+  }
+
+  StoreNode* s1 = this;
+  StoreNode* s2 = def->as_Store();
+  assert(s1->memory_size() == s2->memory_size(), "same size");
+
+  // Check ctrl compatibility
+  Node* ctrl_s1 = s1->in(MemNode::Control);
+  Node* ctrl_s2 = s2->in(MemNode::Control);
+  if (ctrl_s1 != ctrl_s2) {
+    // See if we can bypass a RangeCheck
+    if (!ctrl_s1->is_IfProj() ||
+        !ctrl_s1->in(0)->is_RangeCheck() ||
+        ctrl_s1->in(0)->outcnt() != 2) {
+      return nullptr;
+    }
+    ProjNode* other_proj = ctrl_s1->as_IfProj()->other_if_proj();
+    if (other_proj->is_uncommon_trap_proj(Deoptimization::Reason_range_check) == nullptr ||
```

Comment:
This could be a range check for an unrelated array I suppose. Does it matter?

Reply:
I don't think it matters, no. Do you see a scenario where it would matter?

My argument:
It is safe to do the stores after the RC rather than before it. And if the RC trap relies on the memory state of the stores that were before the RC, then those stores simply don't lose all their uses, and stay in the graph.
After all, we only remove the "last" store by replacing it with the merged store, so the other stores only disappear if they have no other use.

Reply:
> Is there a chance then that we store to the same element twice ... Would it be legal wrt the java specs?

AFAIU, introducing writes that do not exist in original program is an easy way to break JMM conformance. If we merge the writes, we have to make sure the old writes are not done. You _need_ to run jcstress on this change, at very least.